exchanges/order/limits.go

```diff
@@ -0,0 +1,378 @@
+	// ErrMarketAmountExceedsStep is when the amount is not divisible by its
+	// step for a market order
+	ErrMarketAmountExceedsStep = errors.New("market order amount exceeds step limit")
+
+	errCannotValidateAsset         = errors.New("cannot check limit, asset not loaded")
+	errCannotValidateBaseCurrency  = errors.New("cannot check limit, base currency not loaded")
+	errCannotValidateQuoteCurrency = errors.New("cannot check limit, quote currency not loaded")
+	errExchangeLimitAsset          = errors.New("exchange limits not found for asset")
+	errExchangeLimitBase           = errors.New("exchange limits not found for base currency")
+	errExchangeLimitQuote          = errors.New("exchange limits not found for quote currency")
+	errCannotLoadLimit             = errors.New("cannot load limit, levels not supplied")
+	errInvalidPriceLevels          = errors.New("invalid price levels, cannot load limits")
+	errInvalidAmountLevels         = errors.New("invalid amount levels, cannot load limits")
+)
+
+// ExecutionLimits defines minimum and maximum values in relation to
+// order size, order pricing, total notional values, total maximum orders etc
+// for execution on an exchange.
+type ExecutionLimits struct {
+	m   map[asset.Item]map[currency.Code]map[currency.Code]*Limits
+	mtx sync.RWMutex
+}
+
+// MinMaxLevel defines the minimum and maximum parameters for a currency pair
+// for outbound exchange execution
+type MinMaxLevel struct {
+	Pair                currency.Pair
+	Asset               asset.Item
+	MinPrice            float64
+	MaxPrice            float64
+	StepPrice           float64
+	MultiplierUp        float64
+	MultiplierDown      float64
+	MultiplierDecimal   float64
+	AveragePriceMinutes int64
+	MinAmount           float64
+	MaxAmount           float64
+	StepAmount          float64
+	MinNotional         float64
+	MaxIcebergParts     int64
+	MarketMinQty        float64
+	MarketMaxQty        float64
+	MarketStepSize      float64
+	MaxTotalOrders      int64
+	MaxAlgoOrders       int64
+}
+
+// LoadLimits loads all limits levels into memory
+func (e *ExecutionLimits) LoadLimits(levels []MinMaxLevel) error {
+	if len(levels) == 0 {
+		return errCannotLoadLimit
+	}
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+	if e.m == nil {
+		e.m = make(map[asset.Item]map[currency.Code]map[currency.Code]*Limits)
+	}
+
+	for x := range levels {
+		m1, ok := e.m[levels[x].Asset]
+		if !ok {
+			m1 = make(map[currency.Code]map[currency.Code]*Limits)
+			e.m[levels[x].Asset] = m1
+		}
+
+		m2, ok := m1[levels[x].Pair.Base]
+		if !ok {
+			m2 = make(map[currency.Code]*Limits)
+			m1[levels[x].Pair.Base] = m2
+		}
+
+		limit, ok := m2[levels[x].Pair.Quote]
+		if !ok {
+			limit = new(Limits)
+			m2[levels[x].Pair.Quote] = limit
+		}
+
+		if levels[x].MinPrice >= levels[x].MaxPrice {
+			return fmt.Errorf("%w for %s %s supplied min: %f max: %f",
+				errInvalidPriceLevels,
+				levels[x].Asset,
+				levels[x].Pair,
+				levels[x].MinPrice,
+				levels[x].MaxPrice)
```

Comment:
Is this a valid check when min/max price is 0?
> Any of the above variables can be set to 0, which disables that rule in the price filter. In order to pass the price filter, the following must be true for price/stopPrice of the enabled rules:

[here](https://binance-docs.github.io/apidocs/delivery/en/#filters)

It would seem that the rest of the checks are still valid if min max price are 0

Reply:
Yeah, good spotting, can just check if they are set around the wrong way.